classes/statistics/extStatistics.sc

```diff
@@ -480,6 +480,75 @@ oscorr(x,y);
 		};
 		^p
 	}
+
+	// Simple Linear Regression (Least Squares) - Linear Fit
+	linearFit { |x|
+		var size, yFloat, xFloat;
+		var xMean, yMean, sSxy, sSxx, b0, b1;
+
+		size = this.size;
+
+		// cast as Float for precision
+		yFloat = (this.class != Signal).if({
+			this.asFloat
+		}, {
+			this    // Signal elements already Floats. May want to collect as Array(?)
+		});
+		xFloat = (x.isNil).if({
+			Array.series(size, 0.0)
+		}, {
+			x.asFloat  // NOTE: we're not catching x as Signal
+		});
+
+		// mean
+		xMean = xFloat.mean;
+		yMean = yFloat.mean;
+
+		// cross-deviation and deviation about x
+		sSxy = (xFloat * yFloat).sum - (size * xMean * yMean);
+		sSxx = (xFloat.squared).sum - (size * xMean.squared);
+
+		// coefficients
+		b1 = sSxy / sSxx;
+		b0 = yMean - (b1 * xMean);
+
+		^[b0, b1]
+	}
+
+	// Theil-Sen Linear Regression - Linear Fit
+	theilSenFit { |x|
+		var size, yFloat, xFloat;
+		var tuples, slopes, b0, b1;
+
+		size = this.size;
+
+		// cast as Float for precision
+		yFloat = (this.class != Signal).if({
+			this.asFloat
+		}, {
+			this    // Signal elements already Floats. May want to collect as Array(?)
+		});
+		xFloat = (x.isNil).if({
+			Array.series(size, 0.0)
+		}, {
+			x.asFloat  // NOTE: we're not catching x as Signal
```

Comment:
I have a hard time understanding this comment ("NOTE: we're not catching x as Signal"), is it important?

Reply:
I've just thrown this comment in as a consequence of these two methods being extensions to `Collection` rather than `Array`.

Both are inherited by `Signal`, which is very useful... especially given my application of envelope estimation.

For usual use, the user would/should not optionally assign a `Signal` to the `x` arg. This comment in the code is just a little reminder to me that `Signal.asFloat` throws an error.

E.g., `Signal.fill(4, { 1.0 }).asFloat`:

Returns:

> ERROR: Primitive '_AsFloat' failed.
> Failed.
> RECEIVER:
> Instance of Signal {    (0x7f9c5e5dc0a8, gc=90, fmt=03, flg=00, set=02)
>   indexed slots [4]
>       0 : 1
>       1 : 1
>       2 : 1
>       3 : 1
> }
> 
> CALL STACK:
> 	MethodError:reportError
> 		arg this = <instance of PrimitiveFailedError>
> 	Nil:handleError
> 		arg this = nil
> 		arg error = <instance of PrimitiveFailedError>
> 	Thread:handleError
> 		arg this = <instance of Thread>
> 		arg error = <instance of PrimitiveFailedError>
> 	Object:throw
> 		arg this = <instance of PrimitiveFailedError>
> 	Object:primitiveFailed
> 		arg this = <instance of Signal>
> 	Interpreter:interpretPrintCmdLine
> 		arg this = <instance of Interpreter>
> 		var res = nil
> 		var func = <instance of Function>
> 		var code = "Signal.fill(4, { 1.0 }).asFloat"
> 		var doc = nil
> 		var ideClass = <instance of Meta_ScIDE>
> 	Process:interpretPrintCmdLine
> 		arg this = <instance of Main>
> ^^ The preceding error dump is for ERROR: Primitive '_AsFloat' failed.
> Failed.
> RECEIVER: Signal[ 1.0, 1.0, 1.0, 1.0 ]

The other option would be to make these two methods extensions to `Array`, instead. Doing so would mean that to fit data found in a `Signal`, the user would deliberately need to recast as an `Array`. The use pattern would need to be slightly more verbose:

`mySignal.as(Array).theilSenFit`

I'm fine for moving these to `Array`, if you think that's a better way to do things. I had just figured it made some sense to put these in extStatistics, in proximity with the various statistics related methods.

Reply:
I'm just lurking over this PR - very interesting. My 2¢ - from an outsider's perspective - a clearer note would read something like `// NOTE: x may not be a Signal`.

Reply:
Probably we should unify the interface better between Signal and SequenceableCollection. A simple solution in this case is not to call `asFloat` on `x`, but its elements: `x.collect { |item| item.asFloat }`. 

Notwithstanding, I would somehow prefer to avoid implicit conversions and instead make the user aware of the precision issue – it comes up in many other places as well, and it may be even a bit confusing to introduce an convenience that one cannot rely on in other cases. 

Initially, I thought: "it is nice that you can do a linearFit on nested arrays, this is something you can't do if you convert it." But in fact you can: `[[1, 2, 3], 4].asFloat`. So I don't mind that very much.

Reply:
> I'm just lurking over this PR - very interesting. My 2¢ - from an outsider's perspective - a clearer note would read something like `// NOTE: x may not be a Signal`.

@dyfer, agreed, this would be clearer...

Reply:
> Initially, I thought: "it is nice that you can do a linearFit on nested arrays, this is something you can't do if you convert it." But in fact you can: [[1, 2, 3], 4].asFloat. So I don't mind that very much.

I've had a quick look on the possibility of nesting. The algorithm as written doesn't return the expect results... so there'd need to be a re-write... and, a decision as to how far deep to support.

Anyway, I've just added another [commit](c3281b26ea2341c60476fa7b5e95e3d99623a565), clarifying comments. If we're ok with the use of `asFloat`, does this look good to go?